saw-remote-api/src/SAWServer/JVMCrucibleSetup.hs

```diff
@@ -0,0 +1,197 @@
+    go (SetupReturn v) = get >>= \env -> lift $ getSetupVal env v >>= jvm_return bic opts
+    -- TODO: do we really want two names here?
+    go (SetupFresh name@(ServerName n) debugName ty) =
+      do t <- lift $ jvm_fresh_var bic opts (T.unpack debugName) ty
+         (env, cenv) <- get
+         put (env, CEnv.bindTypedTerm (mkIdent n, t) cenv)
+         save name (Val (MS.SetupTerm t))
+    go (SetupAlloc name ty _ (Just _)) =
+      error "attempted to allocate a Java object with alignment information"
+    go (SetupAlloc name (JavaArray n ty) True Nothing) =
+      lift (jvm_alloc_array bic opts n ty) >>= save name . Val
+    go (SetupAlloc name (JavaClass c) True Nothing) =
+      lift (jvm_alloc_object bic opts c) >>= save name . Val
+    go (SetupAlloc name ty False Nothing) =
+      error $ "cannot allocate type: " ++ show ty
+    go (SetupPointsTo src tgt) = get >>= \env -> lift $
+      do ptr <- getSetupVal env src
+         tgt' <- getSetupVal env tgt
+         error "nyi: points-to"
+    go (SetupExecuteFunction args) =
+      get >>= \env ->
+      lift $ traverse (getSetupVal env) args >>= jvm_execute_func bic opts
+    go (SetupPrecond p) = get >>= \env -> lift $
+      getTypedTerm env p >>= jvm_precond
+    go (SetupPostcond p) = get >>= \env -> lift $
+      getTypedTerm env p >>= jvm_postcond
+
+    opts = defaultOptions
+    save name val = modify' (\(env, cenv) -> (Map.insert name val env, cenv))
+
+    getSetupVal ::
+      (Map ServerName ServerSetupVal, CryptolEnv) ->
+      CrucibleSetupVal (P.Expr P.PName) ->
+      JVMSetupM (MS.SetupValue CJ.JVM)
+    getSetupVal _ NullValue = JVMSetupM $ return $ MS.SetupNull ()
+                              {-
+    getSetupVal env (ArrayValue elts) =
+      do elts' <- mapM (getSetupVal env) elts
+         JVMSetupM $ return $ MS.SetupArray () elts'
+    getSetupVal env (FieldLValue base fld) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupField () base' fld
+    getSetupVal env (ElementLValue base idx) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupElem () base' idx
+    getSetupVal _ (GlobalInitializer name) =
+      JVMSetupM $ return $ MS.SetupGlobalInitializer () name
+    getSetupVal _ (GlobalLValue name) =
+      JVMSetupM $ return $ MS.SetupGlobal () name
+         -}
+    getSetupVal (env, _) (ServerValue n) = JVMSetupM $
+      resolve env n >>=
+      \case
+        Val x -> return x -- TODO add cases for the server values that
+                          -- are not coming from the setup monad
+                          -- (e.g. surrounding context)
+    getSetupVal (_, cenv) (CryptolExpr expr) = JVMSetupM $
+      do res <- liftIO $ getTypedTermOfCExp fileReader (biSharedContext bic) cenv expr
+         -- TODO: add warnings (snd res)
+         case fst res of
+           Right (t, _) -> return (MS.SetupTerm t)
+           Left err -> error $ "Cryptol error: " ++ show err -- TODO: report properly
```

Comment:
Presumably "report properly" means to throw a JSON-RPC exception. This seems like something that could be a user mistake rather than a server developer's mistake.

Reply:
I think so, though it'll require some refactoring. I created #85 to track that.